Minimized File Fails to Compile

The following logs are provided whilst compiling Specimin's minimized files in ASHE. We are receiving a **100% failure rate** from what I have experience in bceul-util.

`--args=--outputDirectory "my/temp/dir" --root "my/root/bcel-util/src/main/java" --targetFile "org/plumelib/bcelutil/NoConstraintsVisitor.java" --targetMethod "org.plumelib.bcelutil.NoConstraintsVisitor#setFrame(Frame)"`

15:11:55.915 [main] INFO  edu.njit.jerse.ashe.services.SpeciminTool - 
Specimin output:

> Configure project :
No explicit dependency on the Checker Framework found, using default version 3.42.0

> Task :createCheckerFrameworkManifest UP-TO-DATE
> Task :spotlessInternalRegisterDependencies UP-TO-DATE
> Task :spotlessJava UP-TO-DATE
> Task :spotlessJavaApply UP-TO-DATE
> Task :spotlessMisc UP-TO-DATE
> Task :spotlessMiscApply UP-TO-DATE
> Task :spotlessApply UP-TO-DATE
> Task :compileJava UP-TO-DATE
> Task :processResources UP-TO-DATE
> Task :classes UP-TO-DATE
> Task :run

BUILD SUCCESSFUL in 1s
9 actionable tasks: 1 executed, 8 up-to-date

15:11:55.916 [main] INFO  edu.njit.jerse.ashe.utils.JavaCodeCorrector - Target file minimized successfully.
15:11:55.916 [main] INFO  edu.njit.jerse.ashe.services.SpeciminTool - Compiling minimized file: /my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java
15:11:56.187 [main] INFO  edu.njit.jerse.ashe.services.SpeciminTool - 
Specimin output:
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:3: error: package org.apache.bcel.verifier.structurals does not exist
import org.apache.bcel.verifier.structurals.Frame;
                                           ^
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:4: error: package org.apache.bcel.verifier.structurals does not exist
import org.apache.bcel.verifier.structurals.InstConstraintVisitor;
                                           ^
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:6: error: cannot find symbol
public class NoConstraintsVisitor extends InstConstraintVisitor {
                                          ^
  symbol: class InstConstraintVisitor
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:9: error: cannot find symbol
    public void setFrame(Frame f) {
                         ^
  symbol:   class Frame
  location: class NoConstraintsVisitor
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:8: error: method does not override or implement a method from a supertype
    @Override
    ^
5 errors

15:11:56.187 [main] ERROR edu.njit.jerse.ashe.services.SpeciminTool - Minimized file failed to compile. Exit code: 1
15:11:56.188 [main] INFO  edu.njit.jerse.ashe.Ashe - Skipping...

--- a/src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java
+++ b/src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java
@@ -187,78 +187,51 @@ public class UnsolvedClassOrInterface {
    * @param method the method to be added
    */
   public void addMethod(UnsolvedMethod method) {
-    // Ensure that the method's parameters are all fully-qualified by
-    // marking anything that isn't fully-qualified and isn't a primitive or
-    // in java.lang with the package name of this class.
-    List<String> parameterList = method.getParameterList();
-    for (int i = 0; i < parameterList.size(); i++) {
-      String oldParameterType = parameterList.get(i);
-      String newParameterType = makeParameterTypeFullyQualified(oldParameterType);
-      if (!oldParameterType.equals(newParameterType)) {
-        // This doesn't shift the indices of anything else,
-        // so it won't throw a concurrent modification exception.
-        parameterList.remove(i);
-        parameterList.add(i, newParameterType);
+    // Check for another method with the same parameter list, but with differences in
+    // whether the parameter names are fully-qualified or simple names.
+    List<String> paramList = method.getParameterList();
+    UnsolvedMethod matchingMethod = null;
+    boolean preferOther = true;
+    methods:
+    for (UnsolvedMethod otherMethod : this.methods) {
+      // Skip methods that are definitely different.
+      if (!method.getName().equals(otherMethod.getName())
+          || otherMethod.getParameterList().size() != paramList.size()) {
+        continue;
+      }
+
+      for (int i = 0; i < paramList.size(); ++i) {
+        String paramType = paramList.get(i);
+        String otherParamType = otherMethod.getParameterList().get(i);
+        if ((this.packageName + "." + otherParamType).equals(paramType)) {
+          // In this case, the current method has the FQNs.
+          preferOther = false;
+        } else if (otherParamType.equals(this.packageName + "." + paramType)) {
+          // The other method already has the FQNs, so do nothing here.
+        } else {
+          // if there is ever a difference, skip to the next method; there is
+          // no need to check for exact equality, because methods is a set
+          // so any duplicates won't be added.
+          continue methods;
+        }
+      }
+      matchingMethod = otherMethod;
+    }
+
+    if (matchingMethod != null) {
+      if (preferOther) {
+        // Nothing more to do: the correct method is already present,
+        // and this one is a duplicate with simple names.
+        return;
+      } else {
+        // We need to replace the current variant of the method with this one.
+        // So, remove the current one (the add call below will take care of
+        // adding this method, just as if this was a totally new method).
+        this.methods.remove(matchingMethod);
       }
     }
 
     this.methods.add(method);
-  }
-
-  /**
-   * Converts any non-fully-qualified type names in the parameter type name into fully-qualified
-   * names using the current class' package name.
-   *
-   * @param parameterType the type of the parameter (possibly including type parameters)
-   * @return the same type, but with non-FQNs (besides those in java.lang) converted to FQNs
-   */
-  private String makeParameterTypeFullyQualified(String parameterType) {
-    int typeParamIndex = parameterType.indexOf('<');
-    if (typeParamIndex != -1) {
-      // exclude the <>
-      String typeParams = parameterType.substring(typeParamIndex + 1, parameterType.length() - 1);
-      String baseType = parameterType.substring(0, typeParamIndex);
-      return makeParameterTypeFullyQualified(baseType)
-          + "<"
-          + makeParameterTypeFullyQualified(typeParams)
-          + ">";
-    }
-    // at this point, we're guaranteed to either have a type or a type parameter clause
-    if ("?".equals(parameterType)) {
-      return "?";
-    }
-
-    // handle type parameter bounds
-    String[] extendsAndSuper = {" extends ", " super "};
-    for (String boundType : extendsAndSuper) {
-      if (parameterType.contains(boundType)) {
-        @SuppressWarnings("StringSplitter") // We want the java.lang.String behavior here:
-        // the strings on either side of each " extends " or " super" are guaranteed to be
-        // non-empty.
-        String[] extended = parameterType.split(boundType);
-        StringBuilder result = new StringBuilder();
-        for (int i = 0; i < extended.length; ++i) {
-          result.append(makeParameterTypeFullyQualified(extended[i]));
-          if (i != extended.length - 1) {
-            result.append(boundType);
-          }
-          return result.toString();
-        }
-      }
-    }
-
-    // At this point we know that this is a type name, either simple or fully-qualified.
-    // If the parameter's type is fully-qualified, we're good to go.
-    if (parameterType.indexOf('.') != -1) {
-      return parameterType;
-    }
-    // Otherwise, check if it's a primitive or java.lang class.
-    if (JavaLangUtils.isJavaLangOrPrimitiveName(parameterType)) {
-      return parameterType;
-    }
-    // Finally, if none of the above apply, then add the current class' package name
-    // to make this an FQN.
-    return this.packageName + "." + parameterType;
   }
 
   /**

--- a/src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java
+++ b/src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java
@@ -187,7 +187,78 @@ public class UnsolvedClassOrInterface {
    * @param method the method to be added
    */
   public void addMethod(UnsolvedMethod method) {
+    // Ensure that the method's parameters are all fully-qualified by
+    // marking anything that isn't fully-qualified and isn't a primitive or
+    // in java.lang with the package name of this class.
+    List<String> parameterList = method.getParameterList();
+    for (int i = 0; i < parameterList.size(); i++) {
+      String oldParameterType = parameterList.get(i);
+      String newParameterType = makeParameterTypeFullyQualified(oldParameterType);
+      if (!oldParameterType.equals(newParameterType)) {
+        // This doesn't shift the indices of anything else,
+        // so it won't throw a concurrent modification exception.
+        parameterList.remove(i);
+        parameterList.add(i, newParameterType);
+      }
+    }
+
     this.methods.add(method);
+  }
+
+  /**
+   * Converts any non-fully-qualified type names in the parameter type name into fully-qualified
+   * names using the current class' package name.
+   *
+   * @param parameterType the type of the parameter (possibly including type parameters)
+   * @return the same type, but with non-FQNs (besides those in java.lang) converted to FQNs
+   */
+  private String makeParameterTypeFullyQualified(String parameterType) {
+    int typeParamIndex = parameterType.indexOf('<');
+    if (typeParamIndex != -1) {
+      // exclude the <>
+      String typeParams = parameterType.substring(typeParamIndex + 1, parameterType.length() - 1);
+      String baseType = parameterType.substring(0, typeParamIndex);
+      return makeParameterTypeFullyQualified(baseType)
+          + "<"
+          + makeParameterTypeFullyQualified(typeParams)
+          + ">";
+    }
+    // at this point, we're guaranteed to either have a type or a type parameter clause
+    if ("?".equals(parameterType)) {
+      return "?";
+    }
+
+    // handle type parameter bounds
+    String[] extendsAndSuper = {" extends ", " super "};
+    for (String boundType : extendsAndSuper) {
+      if (parameterType.contains(boundType)) {
+        @SuppressWarnings("StringSplitter") // We want the java.lang.String behavior here:
+        // the strings on either side of each " extends " or " super" are guaranteed to be
+        // non-empty.
+        String[] extended = parameterType.split(boundType);
+        StringBuilder result = new StringBuilder();
+        for (int i = 0; i < extended.length; ++i) {
+          result.append(makeParameterTypeFullyQualified(extended[i]));
+          if (i != extended.length - 1) {
+            result.append(boundType);
+          }
+          return result.toString();
+        }
+      }
+    }
+
+    // At this point we know that this is a type name, either simple or fully-qualified.
+    // If the parameter's type is fully-qualified, we're good to go.
+    if (parameterType.indexOf('.') != -1) {
+      return parameterType;
+    }
+    // Otherwise, check if it's a primitive or java.lang class.
+    if (JavaLangUtils.isJavaLangOrPrimitiveName(parameterType)) {
+      return parameterType;
+    }
+    // Finally, if none of the above apply, then add the current class' package name
+    // to make this an FQN.
+    return this.packageName + "." + parameterType;
   }
 
   /**

--- a/src/main/java/org/checkerframework/specimin/UnsolvedMethod.java
+++ b/src/main/java/org/checkerframework/specimin/UnsolvedMethod.java
@@ -1,5 +1,6 @@
 package org.checkerframework.specimin;
 
+import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 import org.checkerframework.checker.nullness.qual.Nullable;
@@ -22,7 +23,7 @@ public class UnsolvedMethod {
    * The list of the types of the parameters of the method. (Right now we won't touch it until the
    * new variant of SymbolSolver is available)
    */
-  private List<String> parameterList;
+  private final List<String> parameterList;
 
   /** This field is set to true if this method is a static method */
   private boolean isStatic = false;
@@ -113,13 +114,12 @@ public class UnsolvedMethod {
   }
 
   /**
-   * Getter for the parameter list. Caution: this method returns an actual pointer to the parameter
-   * list, so updates to this list will be reflected in the method.
+   * Getter for the parameter list. Note that the list is read-only.
    *
    * @return the parameter list
    */
   public List<String> getParameterList() {
-    return parameterList;
+    return Collections.unmodifiableList(parameterList);
   }
 
   /** Set isStatic to true */

--- a/src/main/java/org/checkerframework/specimin/UnsolvedMethod.java
+++ b/src/main/java/org/checkerframework/specimin/UnsolvedMethod.java
@@ -112,6 +112,16 @@ public class UnsolvedMethod {
     return name;
   }
 
+  /**
+   * Getter for the parameter list. Caution: this method returns an actual pointer to the parameter
+   * list, so updates to this list will be reflected in the method.
+   *
+   * @return the parameter list
+   */
+  public List<String> getParameterList() {
+    return parameterList;
+  }
+
   /** Set isStatic to true */
   public void setStatic() {
     isStatic = true;

--- a/src/test/java/org/checkerframework/specimin/OverrideDoubleTest.java
+++ b/src/test/java/org/checkerframework/specimin/OverrideDoubleTest.java
@@ -1,0 +1,15 @@
+package org.checkerframework.specimin;
+
+import java.io.IOException;
+import org.junit.Test;
+
+/** The test for <a href="https://github.com/kelloggm/specimin/issues/185">...</a>. */
+public class OverrideDoubleTest {
+  @Test
+  public void runTest() throws IOException {
+    SpeciminTestExecutor.runTestWithoutJarPaths(
+        "overridedouble",
+        new String[] {"com/example/Simple.java"},
+        new String[] {"com.example.Simple#setFrame(Frame)"});
+  }
+}

--- a/src/test/resources/overridedouble/expected/com/example/Simple.java
+++ b/src/test/resources/overridedouble/expected/com/example/Simple.java
@@ -1,0 +1,10 @@
+package com.example;
+
+import org.example.Frame;
+import org.example.InstConstraintVisitor;
+
+public class Simple extends InstConstraintVisitor {
+
+    public void setFrame(Frame f) {
+    }
+}

--- a/src/test/resources/overridedouble/expected/org/example/Frame.java
+++ b/src/test/resources/overridedouble/expected/org/example/Frame.java
@@ -1,0 +1,4 @@
+package org.example;
+
+public class Frame {
+}

--- a/src/test/resources/overridedouble/expected/org/example/InstConstraintVisitor.java
+++ b/src/test/resources/overridedouble/expected/org/example/InstConstraintVisitor.java
@@ -1,0 +1,8 @@
+package org.example;
+
+public class InstConstraintVisitor {
+
+    public void setFrame(org.apache.bcel.verifier.structurals.Frame parameter0) {
+        throw new Error();
+    }
+}

--- a/src/test/resources/overridedouble/expected/org/example/InstConstraintVisitor.java
+++ b/src/test/resources/overridedouble/expected/org/example/InstConstraintVisitor.java
@@ -2,7 +2,7 @@ package org.example;
 
 public class InstConstraintVisitor {
 
-    public void setFrame(org.apache.bcel.verifier.structurals.Frame parameter0) {
+    public void setFrame(org.example.Frame parameter0) {
         throw new Error();
     }
 }

--- a/src/test/resources/overridedouble/input/com/example/Simple.java
+++ b/src/test/resources/overridedouble/input/com/example/Simple.java
@@ -1,0 +1,16 @@
+package com.example;
+
+import org.example.Frame;
+import org.example.InstConstraintVisitor;
+
+public class Simple extends InstConstraintVisitor {
+
+    // Target method. The name of this test comes from the fact
+    // that this method triggered Specimin to accidentally create
+    // two copies of this method in InstConstraintVisitor: one
+    // with the fully-qualified parameter name, and one without.
+    // This is the test for https://github.com/kelloggm/specimin/issues/185.
+    @Override
+    public void setFrame(Frame f) {
+    }
+}

